crates/turborepo-repository/src/package_manager/mod.rs

```diff
@@ -483,7 +484,52 @@ impl PackageManager {
             &globs.validated_exclusions,
             globwalk::WalkType::Files,
         )?;
-        Ok(files.into_iter())
+
+        // we need to remove package.json files that are in subfolders of others so that
+        // we don't yield subpackages. sort, keep track of the parent of last
+        // json we encountered, and only yield it if it's not a subfolder of it
+        //
+        // ideally we would do this during traversal, but walkdir doesn't support
+        // inorder traversal so we can't
+        let mut last_parent = None;
+        Ok(files
+            .into_iter()
+            .sorted_by(|a, b| {
+                // get an iterator of the components of each path, and zip them together
+                let mut segments = a.components().is_last().zip(b.components().is_last());
+
+                // find the first pair of components that are different, and compare them
+                // if one of the segments is the last, then the other is a subfolder of it.
+                // we must always yield 'file-likes' (the last segment of a path) ahead of
+                // subfolders (non-last segments) so that we can guarantee we find the
+                // package.json before processing its subfolders
+                segments
+                    .find_map(|((a_last, a_cmp), (b_last, b_cmp))| {
+                        if a_last == b_last {
+                            match a_cmp.cmp(&b_cmp) {
+                                std::cmp::Ordering::Equal => None,
+                                other => Some(other),
+                            }
+                        } else if a_last {
+                            Some(std::cmp::Ordering::Less)
+                        } else {
+                            Some(std::cmp::Ordering::Greater)
+                        }
+                    })
+                    .unwrap_or(std::cmp::Ordering::Equal)
+            })
+            .filter(move |entry| {
+                match &last_parent {
+                    // last_parent is the parent of the last json we yielded. if the current
+                    // entry is a subfolder of it, we don't want to yield it
+                    Some(parent) if entry.starts_with(parent) => false,
+                    // update last_parent to the parent of the current entry
+                    _ => {
+                        last_parent = Some(entry.parent().unwrap().to_owned());
+                        true
+                    }
+                }
+            }))
```

Comment:
This filtering code feels like a prime candidate to be pulled out into a helper function and have a bunch of unit tests written against it.